distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/common/downloads/AsynchronousClientDownloader.java

```diff
@@ -0,0 +1,93 @@
+ * is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+package ddf.test.itests.common.downloads;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.is;
+
+import java.io.IOException;
+import java.nio.charset.Charset;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+
+import org.apache.commons.io.IOUtils;
+import org.apache.http.HttpResponse;
+import org.apache.http.client.methods.HttpGet;
+import org.apache.http.message.BasicHttpResponse;
+
+/**
+ * Provides an asynchronous handle to the asynchronous request and synchronous retrieval of a
+ * product utilizing both endpoints in the process.
+ */
+class AsynchronousClientDownloader extends ClientDownloader {
+    private Future downloadRequestFuture = null;
+
+    private String metacardId = null;
+
+    @Override
+    public void startDownload(String metacardId) {
+        client.start();
+        this.metacardId = metacardId;
+        downloadRequestFuture = client.execute(new HttpGet(ClientDownloader.getAsyncDownloadUrl(
+                metacardId)), null);
+        try {
+            HttpResponse response = (BasicHttpResponse) downloadRequestFuture.get();
+            int responseCode = response.getStatusLine()
+                    .getStatusCode();
+            assertThat("Asynchronous request was not successful (responseCode != 200)",
+                    responseCode,
+                    is(200));
+        } catch (InterruptedException e) {
+            Thread.currentThread()
+                    .interrupt();
+        } catch (ExecutionException e) {
+            throw new RuntimeException("Exception occurred in future's thread", e);
+        }
+
+        if (downloadRequestFuture.isCancelled()) {
+            throw new IllegalStateException("Future was cancelled");
+        }
+        if (!downloadRequestFuture.isDone()) {
+            throw new IllegalStateException("Future has not finished");
+        }
+    }
+
+    @Override
+    public void stopDownload() {
+        // TODO: [DDF-2287] Cannot implement until async cancel is available
+    }
+
+    @Override
+    protected String processResults() throws ExecutionException {
+        if (downloadRequestFuture == null) {
+            throw new IllegalStateException("Download was never started, so no future exists");
+        }
+
+        Future future =
+                client.execute(new HttpGet(ClientDownloader.getSyncDownloadUrl(this.metacardId)),
+                        null);
+
+        try {
+            HttpResponse response = (BasicHttpResponse) future.get();
+            return IOUtils.toString(response.getEntity()
+                    .getContent(), "UTF-8");
+        } catch (InterruptedException e) {
+            Thread.currentThread()
+                    .interrupt();
+            return processResults();
```

Comment:
You should stop processing and throw an ExecutionException.  Most likely if you are being interrupted, the system is trying to shutdown.  We do not want to hang the system shutdown on a large download.

http://www.ibm.com/developerworks/library/j-jtp05236/index.html